### PR TITLE
feat: unlimited candidates

### PR DESCRIPTION
**Pull request by weichweich**
## fixes KILTprotocol/srlabs-audit/issues/2

* remove 'cache' for selected candidates, calculate on the fly
* replace {increase, decrease}_max_candidate_stake with simple set_max_candidate_stake
  * don't adjust existing collator stake. If root wants to adjust the stake, they need to adjust the existing stake manually with root calls. 

### todo

- [ ] storage migration: remove `selected_candidate` list
- [ ] storage migration: initialize `CandidateCount`
- [x] Events: enter & leave top candidates
- [ ] optimization: don't iter over collators to update total stake

### custom types

This PR introduces new custom JS-types which are required for compatibility with [our SDK](https://github.com/KILTprotocol/sdk-js) and the [Polkadot Apps](https://polkadot.js.org/apps/#/extrinsics). Please use the following types to test the code with the Polkadot Apps:

<details>
  <summary>JS-Types</summary>
  
  ```js
{
  Candidate: {
    id: "AccountId",
    stake: "Balance",

    delegators: "Vec<Stake>",
    total: "Balance",
    state: "CollatorStatus",
  },
  CandidateOf: "Candidate",
  Collator: undefined,
  CollatorOf: undefined,
}
  ```
</details>

## Checklist:

- [ ] I have verified that the code works
  - [ ] No panics! (checked arithmetic ops, no indexing `array[3]` use `get(3)`, ...)
- [ ] I have verified that the code is easy to understand
  - [ ] If not, I have left a well-balanced amount of inline comments
- [ ] This PR does not introduce new custom types
  - [ ] If not, I have opened a companion PR with the type changes in the [KILT types-definitions repository](https://github.com/KILTprotocol/type-definitions/pulls)
- [ ] I have [left the code in a better state](https://deviq.com/principles/boy-scout-rule)
- [ ] I have documented the changes (where applicable)

### CODE REVIEW

File: pallets/parachain-staking/src/set.rs [hunk 1]
```diff
@@ -55,54 +56,71 @@ impl<T: Ord + Clone, S: Get<u32>> OrderedSet<T, S> {
...
+	///   had to removed.
+	/// * Ok(None) if the element was added without removing an element.
+	/// * Err(true) if the set is full and the new element has a lower rank than
+	///   the lowest element in the set.
+	/// * Err(false) if the element is already in the set.
+	pub fn try_insert_replace(&mut self, value: T) -> Result<Option<T>, bool> {
+		// the highest allowed index
+		let highest_index: usize = S::get().saturating_sub(1).saturated_into();
+		if highest_index.is_zero() {
+			return Err(false);
```
Comment (weichweich) [inline pallets/parachain-staking/src/set.rs:110]: should be `Err(true)`

File: pallets/parachain-staking/src/set.rs [hunk 2]
```diff
@@ -55,54 +56,71 @@ impl<T: Ord + Clone, S: Get<u32>> OrderedSet<T, S> {
...
+	/// * Ok(Some(old_element)) if the new element was added and an old element
+	///   had to removed.
+	/// * Ok(None) if the element was added without removing an element.
+	/// * Err(true) if the set is full and the new element has a lower rank than
+	///   the lowest element in the set.
+	/// * Err(false) if the element is already in the set.
+	pub fn try_insert_replace(&mut self, value: T) -> Result<Option<T>, bool> {
+		// the highest allowed index
+		let highest_index: usize = S::get().saturating_sub(1).saturated_into();
+		if highest_index.is_zero() {
```
Comment (weichweich) [inline pallets/parachain-staking/src/set.rs:109]: should check that `S::get().is_zero()`

File: pallets/parachain-staking/src/lib.rs [hunk 3]
```diff
@@ -1052,35 +1008,20 @@ pub mod pallet {
...
-					owner: acc.clone(),
-					amount: stake,
-				})
-				.map_err(|_| Error::<T>::TooManyCollatorCandidates)?;
-			// should never fail but let's be safe
-			ensure!(insert_candidate, Error::<T>::CandidateExists);
-
-			Self::increase_lock(&acc, stake, BalanceOf::<T>::zero())?;
+			let candidate = Candidate::new(sender.clone(), stake);
+			<CandidatePool<T>>::insert(&sender, candidate);
```
Comment (weichweich) [inline pallets/parachain-staking/src/lib.rs:1014]: Turbo fish!

File: pallets/parachain-staking/src/lib.rs [hunk 4]
```diff
@@ -1231,50 +1173,31 @@ pub mod pallet {
...
+		/// - Writes: Total, CollatorState, TopCandidates, SelectedCandidates
 		/// # </weight>
 		#[pallet::weight(<T as pallet::Config>::WeightInfo::cancel_leave_candidates(
 			T::MaxCollatorCandidates::get(),
 			T::MaxDelegatorsPerCollator::get(),
 		))]
 		pub fn cancel_leave_candidates(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
-			let acc = ensure_signed(origin)?;
-			let mut state = <CollatorState<T>>::get(&acc).ok_or(Error::<T>::CandidateNotFound)?;
+			let collator = ensure_signed(origin)?;
```
Comment (weichweich) [inline pallets/parachain-staking/src/lib.rs:1187]: `sender` it's not a collator

File: pallets/parachain-staking/src/lib.rs [hunk 5]
```diff
@@ -1390,9 +1313,12 @@ pub mod pallet {
 			Self::prep_unstake(&collator, less)?;
 
 			if state.is_active() {
-				Self::update(collator.clone(), state.total)?;
+				Self::update_top_candidates(collator.clone(), before, state.total);
 			}
-			<CollatorState<T>>::insert(&collator, state);
+
+			Self::update_top_candidates(collator.clone(), before, state.total);
```
Comment (weichweich) [inline pallets/parachain-staking/src/lib.rs:1319]: remove this line

File: pallets/parachain-staking/src/lib.rs [hunk 6]
```diff
@@ -2190,23 +2134,25 @@ pub mod pallet {
...
-				stake_to_remove.amount,
-				state.id.clone(),
-				state.total,
-			));
+			if let Some(stake_to_remove) = stake_to_remove {
+				// update total stake
+				state.total = state
+					.total
+					.saturating_sub(stake_to_remove.amount)
+					.saturating_add(stake.amount);
```
Comment (weichweich) [inline pallets/parachain-staking/src/lib.rs:2142]: check if this is fine. What if nobody was removed from the delegators? We don't increase the total stake?

File: pallets/parachain-staking/src/lib.rs (hunk 5 above)
Comment (weichweich) [reply pallets/parachain-staking/src/lib.rs]: fixed in 159533c5b592cf47ea6e84f96dc24e14845eb14d

File: pallets/parachain-staking/src/lib.rs (hunk 3 above)
Comment (weichweich) [reply pallets/parachain-staking/src/lib.rs]: fixed in c28b05d29691cb1dfb56131412e2519bf927d6cb

File: pallets/parachain-staking/src/set.rs [hunk 7]
```diff
@@ -55,54 +56,71 @@ impl<T: Ord + Clone, S: Get<u32>> OrderedSet<T, S> {
...
+			return Err(false);
+		}
+		match self.try_insert(value.clone()) {
+			Err(loc) if loc <= highest_index => {
 				// always replace the last element
+				let last_idx = self.len().saturating_sub(1);
+				// accessing by index wont panic since we checked the index
```
[the remark this answers is not in the document]
Comment (weichweich) [reply pallets/parachain-staking/src/set.rs]: fixed in 35e09b280754520e650dceb8309d0156a20e9384

File: pallets/parachain-staking/src/set.rs (hunk 2 above)
Comment (weichweich) [reply pallets/parachain-staking/src/set.rs]: fixed in 661b80cab9592d6a4124bd10b08d3c8db8c28c61

File: pallets/parachain-staking/src/lib.rs (hunk 4 above)
Comment (weichweich) [reply pallets/parachain-staking/src/lib.rs:1187]: fixed in 6767790a29e9b6a762ebad242cdb7b129a1ff4c8

File: pallets/parachain-staking/src/lib.rs (hunk 6 above)
Comment (weichweich) [reply pallets/parachain-staking/src/lib.rs:2139]: fixed here d404ef222601c6551ee27a7951ed788593eacbca just to be safe

**Review by wischli — APPROVED** [review]
LGTM! Great improvement

**Comment by wischli** [comment]
Please [open a PR for the updated types](https://github.com/KILTprotocol/type-definitions/compare).

**Comment by weichweich** [comment]
* [Types PR is open but untested](https://github.com/KILTprotocol/type-definitions/pull/19)
* benchmarks & migrations will follow in a different PR

**Pull request merged**